Scala3 Documentation (#3)
scalacenter/docs.scala-lang: _overviews/scala3-book/concurrency.md

---
title: Concurrency
type: chapter
description: This page discusses how Scala concurrency works, with an emphasis on Scala Futures.
num: 68
previous-page: ca-summary
next-page: scala-tools
---

> When you want to write parallel and concurrent applications in Scala, you *can* use the native Java `Thread` — but the Scala [Future](https://www.scala-lang.org/api/current/scala/concurrent/Future$.html) makes parallel/concurrent programming much simpler, so it’s preferred, and covered in this chapter.

Reviewer: ```suggestion
When you want to write parallel and concurrent applications in Scala, you *can* use the native Java `Thread` — but the Scala [Future](https://www.scala-lang.org/api/current/scala/concurrent/Future$.html) offers a more high level and idiomatic approach so it’s preferred, and covered in this chapter.
```

Here’s a description of the Scala `Future` from its Scaladoc:

>“A `Future` represents a value which may or may not *currently* be available, but will be available at some point, or an exception if that value could not be made available.” 

To help demonstrate what that means, let’s first look at single-threaded programming. In the single-threaded world you bind the result of a method call to a variable like this:

```scala
def aShortRunningTask(): Int = 42
val x = aShortRunningTask()
```

In this code, the value `42` is immediately bound to the variable `x`.

When you’re working with a `Future`, the assignment process looks similar:

```scala
def aLongRunningTask(): Future[Int] = ???
val x = aLongRunningTask()
```

The main difference in this case is that because `aLongRunningTask` takes an indeterminate amount of time to return, the value in `x` may or may not be *currently* available, but it will be available at some point — in the future.

In this chapter you’ll see how to use futures, including how to run multiple futures in parallel and combine their results in a `for` expression. You’ll also see other methods that are used to handle the value in a future once it returns.

> 

Reviewer: I think it might be useful to reason in blocking terms too:
```scala
def aShortRunningTask(): Int = 42
val x = aShortRunningTask()
println("Here")
```
The print is executed only after the `aShortRunningTask` completes. In the second case, the print can be executed even if the computation is not completed

>An important point to know about futures is that they’re intended as a one-shot, “Handle this relatively slow computation on some other thread, and call me back with a result when you’re done” construct. As a point of contrast, [Akka](https://akka.io) actors are intended to run for a long time and respond to many requests during their lifetime. While an actor may live forever, a future is intended to be run only once.

## An example in the REPL

A future is used to create a temporary pocket of concurrency. For instance, you use a future when you need to call an algorithm that runs an indeterminate amount of time — such as calling a remote microservice — so you want to run it off of the main thread.

To demonstrate how this works, let’s start with a `Future` example in the REPL. First, paste in these required `import` statements:

```scala
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.util.{Failure, Success}
```

Now you’re ready to create a future. For this example, first define a long-running, single-threaded algorithm:

```scala
def longRunningAlgorithm =
  Thread.sleep(10_000)
  42
```

> That fancy algorithm returns the integer value `42` after a ten second delay. Now call that algorithm by passing it into the `Future` constructor, and assigning the result to a variable:

Reviewer: ```suggestion
That fancy algorithm returns the integer value `42` after a ten second delay. Now call that algorithm by wrapping it into the `Future` constructor, and assigning the result to a variable:
```

```scala
scala> val f = Future(longRunningAlgorithm)
f: scala.concurrent.Future[Int] = Future(<not completed>)
```

Right away your future begins running. If you immediately check the value of the variable `f`, you see that the future hasn’t completed yet:

```scala
scala> f
val res1: scala.concurrent.Future[Int] = Future(<not completed>)
```

But if you check again after ten seconds, you’ll see that it completes successfully:

```scala
scala> f
val res2: scala.concurrent.Future[Int] = Future(Success(42))
```

While that’s a relatively simple example, it shows the basic approach: Just construct a new `Future` with your long-running algorithm.

> One thing to notice is that the `42` you expected is wrapped in a `Success`, which is further wrapped in a `Future`. This is a key concept to understand: The value in a `Future` is always an instance of one of the *scala.util.Try* types: `Success` or `Failure`. Therefore, when you work with the result of a future, you use the usual `Try`-handling techniques.

Reviewer: ```suggestion
One thing to notice is that the `42` you expected is wrapped in a `Success`, which is further wrapped in a `Future`. This is a key concept to understand: the value in a `Future` is always an instance of one of the *scala.util.Try* types: `Success` or `Failure`. Therefore, when you work with the result of a future, you use the usual `Try`-handling techniques.
```

<!-- TODO: link to the Try/Success/Failure docs -->

### Using `map` with futures

> `Future` has a `map` method, which you use just like the `map` method on the collections classes. This is what the result looks like when you call `map` right after creating the variable `f`:

Reviewer: ```suggestion
`Future` has a `map` method, which you use just like the `map` method on collections. This is what the result looks like when you call `map` right after creating the variable `f`:
```

- You construct futures to run tasks off of the main thread
- Futures are intended for one-shot, potentially long-running concurrent tasks that *eventually* return a value; they create a temporary pocket of concurrency
- A future starts running as soon as you construct it
- A benefit of futures over threads is that they work with `for` expressions, and come with a variety of callback methods that simplify the process of working with concurrent threads
- When you work with futures you don’t have to concern yourself with the low-level details of thread management
- You handle the result of a future with callback methods like `onComplete` and `andThen`, or transformation methods like `filter`, `map`, etc.
- The value inside a `Future` is always an instance of one of the `Try` types: `Success` or `Failure`
- If you’re using multiple futures to yield a single result, combine them in a `for` expression

Also, as you saw with the `import` statements in these examples, the Scala `Future` depends on an `ExecutionContext`. This relationship, and many more details about futures, are explained in the Reference documentation.

> 

Reviewer: I would include an example of flatmap.
